### ISSUE TITLE
Page: OpenGL Instead of ImageView

### ISSUE THREAD

**Issue by ghost** [mobileappdevhm19/HINT-Reader#43]
@gsocher had a suggestion: Instead of having the backend deliver an image file, have the frontend display an OpenGL es scene, mapping letters as textures onto quads.
Each letter on a page would be its own textured quad.
The backend would be responsible for delivering the GL objects, the frontend would need to implement the code neccessary to display a scene – at least, that's as fa as I understood it.

Prof. Socher kindly made an example project displaying a scene via Flutter, found in this repo:
https://github.com/mobileappdevhm19/opengl_texture_widget_example

What are everybody's opinions on this?

**Comment by wildeKaiserin** [comment]
What about discussing this issue together on Monday at our next meeting?

**Comment by maxiking203** [comment]
We clould try to include this into the current hintreader as a new book

**Comment by ghost** [comment]
@wildeKaiserin absolutely!

@maxiking203 I like that a lot - keeping our options open in case it doesn't work seems sensible. We don't have a LINTer complaining about duplicate code, have we? ^^

**Comment by gsocher** [comment]
Can anyone try to make the architecture of [https://github.com/mobileappdevhm19/opengl_texture_widget_example](https://github.com/mobileappdevhm19/opengl_texture_widget_example)
work such that 
![texture_widget_example](https://camo.githubusercontent.com/6f7491cd882605abe953717e2889a30910cb17b6/68747470733a2f2f63646e2d696d616765732d312e6d656469756d2e636f6d2f6d61782f313630302f312a6166537a71592d7677586d70676d6d6e4149614634772e706e67)
C++ OpenGL Code is not in the platform dependent parts of the code. Please try a very simple OpenGL piece of code of your choice to make a prototype that shows where the shared C++ OpenGL code will be located in the project

**Comment by wildeKaiserin** [comment]
During the holidays, we had no time for any implementation details. We will take a look at it tomorrow (next sprint) together.

**Comment by ghost** [comment]
Issue on ice for now, in favour of https://github.com/mobileappdevhm19/HINT-Reader/issues/48

**Comment by wildeKaiserin** [comment]
I talked to Markus Miller on Wednesday because of this topic. He said that he will have to take a closer look at it.
Update: He sent me an email tonight and told me, that textrendering works now in OpenGL!

**Comment by ghost** [comment]
I'd file that under backup plans, so far we have a working prototype that loads and caches images from a webserver. (see https://github.com/mobileappdevhm19/HINT-Reader/issues/48 )
I'm not sure we have the capacity to keep pursuing this avenue at the same time.

**Comment by ghost** [comment]
Au contraire, we are all in on the openGL backend solution.

For your information, it doesn't seem to compile on iOS right now, because the objC implementation OpenglTexturePlugin.m is missing a header file?
That's just for the record though, we were saying the reference device is Android anyway, and iOS support would be down the line.

**Comment by wildeKaiserin** [comment]
At the moment, it doesn't work on any of our devices... it doesn't compile...

**Comment by maxiking203** [comment]
As I was able run Mr. Ruckerts project yesterday I tried so get our UI with the Bookshelf and the BookView into the new project. Now instead of our images we are displaying the openGL widget in the BookView.
The code for that is located in openGLBook branch.

**Comment by ghost** [comment]
Installation instructions in the readme of the openGLBook branch:

https://github.com/mobileappdevhm19/HINT-Reader/blob/feature/openGLBook/README.md

**Comment by maxiking203** [comment]
Removed "Prio 1" and "help wanted" label from this issue.
 The openGL Project is included in our hintreader and is running (at least on my laptop, and two phones). Where a more or less waiting for updated functions in the backend.

**Comment by GiorgioBullo** [comment]
We should update the README file in the GLM section.
When you download glm-0.9.9.5.7, after the extraction you will have a glm folder. Inside this folder there is another glm folder. The latter is the one that has to be copied in the indicated path.

It's not a big deal, but maybe it is an help for those that like me get panicked when receives a not expected error hhahaha

**Comment by ghost** [comment]
Readme has been updated with clarification regarding the paths. Thank you for spotting that!

**Comment by wildeKaiserin** [comment]
Release Date: **02.07.2019**
We will release the feature branch, not the master branch because backend proves untestable in our current setup

**Comment by wildeKaiserin** [comment]
A new version of the backend is available! We will now try to implement it in the Hint Reader, but we cannot guarantee if it works until friday because we already have some exams.

**Comment by maxiking203** [comment]
For docomentation - Error I get when I try to run the app on AndroidStudio:
Launching lib\main.dart on SM G930F in debug mode...
Initializing gradle...
Resolving dependencies...
Running Gradle task 'assembleDebug'...
Built build\app\outputs\apk\debug\app-debug.apk.
Installing build\app\outputs\apk\app.apk...
D/OpenglTexturePlugin(13348): create {width=800.0, height=600.0}
D/OpenGL.Worker(13348): Creating render thread
V/Surface (13348): sf_framedrop debug : 0x4f4c, game : false, logging : 0
D/mali_winsys(13348): EGLint new_window_surface(egl_winsys_display *, void *, EGLSurface, EGLConfig, egl_winsys_surface **, egl_color_buffer_format *, EGLBoolean) returns 0x3000,  [800x600]-format:1
I/libgl2jni(13348): init (800 px, 600 px)
I/libgl2jni(13348): GL Version = OpenGL ES 3.2 v1.r22p0-01rel0.f294e54ceb2cb2d81039204fa4b0402e
I/libgl2jni(13348): GL Vendor = ARM
I/libgl2jni(13348): GL Renderer = Mali-T880
I/libgl2jni(13348): GL Extensions = GL_EXT_debug_marker GL_ARM_rgba8 GL_ARM_mali_shader_binary GL_OES_depth24 GL_OES_depth_texture GL_OES_depth_texture_cube_map GL_OES_packed_depth_stencil GL_OES_rgb8_rgba8 GL_EXT_read_format_bgra GL_OES_compressed_paletted_texture GL_OES_compressed_ETC1_RGB8_texture GL_OES_standard_derivatives GL_OES_EGL_image GL_OES_EGL_image_external GL_OES_EGL_image_external_essl3 GL_OES_EGL_sync GL_OES_texture_npot GL_OES_vertex_half_float GL_OES_required_internalformat GL_OES_vertex_array_object GL_OES_mapbuffer GL_EXT_texture_format_BGRA8888 GL_EXT_texture_rg GL_EXT_texture_type_2_10_10_10_REV GL_OES_fbo_render_mipmap GL_OES_element_index_uint GL_EXT_shadow_samplers GL_OES_texture_compression_astc GL_KHR_texture_compression_astc_ldr GL_KHR_texture_compression_astc_hdr GL_KHR_texture_compression_astc_sliced_3d GL_KHR_debug GL_EXT_occlusion_query_boolean GL_EXT_disjoint_timer_query GL_EXT_blend_minmax GL_EXT_discard_framebuffer GL_OES_get_program_binary GL_OES_texture_3D GL_EXT_texture_storage GL_EXT_multi
I/libgl2jni(13348): setupGraphics(385.440002 pt, 289.079987 pt) (800 px, 600 px)
E/.\hint.w(13348): Unable to open file /data/data/com.yourcompany.opengltextureexample/files/math.hnt
Syncing files to device SM G930F...
Lost connection to device.
Error writing "build/flutter_assets/AssetManifest.json" to DevFS: HttpException: Connection closed before full header was received, uri = http://127.0.0.1:53648/dq6jPJe9YP8=/
Error writing "build/flutter_assets/LICENSE" to DevFS: HttpException: Connection closed before full header was received, uri = http://127.0.0.1:53648/dq6jPJe9YP8=/
Error writing "build/flutter_assets/FontManifest.json" to DevFS: HttpException: Connection closed before full header was received, uri = http://127.0.0.1:53648/dq6jPJe9YP8=/
Error writing "build/flutter_assets/packages/cupertino_icons/assets/CupertinoIcons.ttf" to DevFS: HttpException: Connection closed before full header was received, uri = http://127.0.0.1:53648/dq6jPJe9YP8=/
Error writing "build/flutter_assets/fonts/MaterialIcons-Regular.ttf" to DevFS: HttpException: Connection closed before full header was received, uri = http://127.0.0.1:53648/dq6jPJe9YP8=/


Tried to store the Hint files everywhere in my phone + (changing the path of "hint_start" in gl_code), doesn't work. The /data/data folder is not accessible on my phone due to missing root-permissions.
So I'm more than over asked, also there is minimal documentation for the new backend.

In conclusion I am not willing to put anymore time to this, especially that close to the exams.
Of course for the backend its huge improvement to "display" real hint-files. For the frontend it doesn't really matter. So as I am not able get the new backend project running by itself I can not include it in our hintreader. In other words the current version the feature/openGLBook branch will be our Release!

**Comment by gsocher** [comment]
My solution on putting math.hnt on my phone:
Add to opengl_texture_widget_example\android\src\main\AndroidManifest.xml
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
This gives the Android the permissions to access the file system.
Then I plugged my phone into my computer and copied math.hnt to /Download/
I put in gl_code.cpp:
hint_start("/storage/emulated/0/Download/math.hnt");

**Comment by maxiking203** [comment]
@gsocher thank you for the hint. 
I will definitely try that. But unfortunately, due to the lack of time, the new backend is and will not be part of the release from our last sprint :(

**Comment by gsocher** [comment]
https://gitlab.lrz.de/mobileappdev19/hint-reader/tree/feature/openGLBook (branch feature/openGLBook) 
the file math.hnt needs to copied to the Download folder on the Android device. Storage permissions need to be given manually.
